### ISSUE TITLE
[DTensor] Local tensor stride incorrect for uneven case

### ISSUE THREAD

**Issue by awgu** [pytorch/pytorch#118836]
For uneven sharding case, the local tensor on rank that does not have the full shard can be wrong, using the full shard's stride instead of accounting for the local tensor missing some elements from the unevenness.

```
import torch
import torch.distributed as dist

from torch.distributed._tensor import init_device_mesh
from torch.distributed.tensor.parallel import parallelize_module, RowwiseParallel


mesh = init_device_mesh(device_type="cuda", mesh_shape=(2,))
lin = torch.nn.Linear(9, 3, device="cuda")
parallelize_module(lin, mesh, RowwiseParallel())
rank = dist.get_rank()
print(
    f"[Rank {rank}] placements={lin.weight._spec.placements} | global shape={lin.weight.shape}\n"
    f"{lin.weight._local_tensor.shape} | {lin.weight._local_tensor.stride()}"
)
```

```
torchrun --standalone --nproc_per_node=2 repro.py
```

```
[Rank 0] placements=(Shard(dim=1),) | global shape=torch.Size([3, 9])
torch.Size([3, 5]) | (5, 1)
[Rank 1] placements=(Shard(dim=1),) | global shape=torch.Size([3, 9])
torch.Size([3, 4]) | (5, 1)
```

On Rank 1, contiguous strides would be `(4 ,1)` instead of `(5, 1)`. For `DTensor`'s user-facing API, this may not matter since the `_local_tensor` is not exposed. However, because in per-parameter FSDP, we were directly using the `_local_tensor` and sharding it further on dim-0 ourselves and using the stride info, we hit a correctness error.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225

**Comment by awgu** [comment]
I worked around this issue by computing the contiguous strides directly and not using the tensor's `.stride()`, so this is not blocking anything. I imagine it might be tricky to special case the uneven local tensors' stride.

**Comment by SherlockNoMad** [comment]
Looks like the issue has been fixed. Rerun the repro and I got 

```
[Rank 1] placements=(Shard(dim=1),) | global shape=torch.Size([3, 9])
torch.Size([3, 4]) | (4, 1)
[Rank 0] placements=(Shard(dim=1),) | global shape=torch.Size([3, 9])
torch.Size([3, 5]) | (5, 1)
```
